vsock/src/vhu_vsock.rs

```diff
@@ -0,0 +1,488 @@
+    UnixConnect(std::io::Error),
+    #[error("Unable to write to unix stream")]
+    UnixWrite,
+    #[error("Unable to push data to local tx buffer")]
+    LocalTxBufFull,
+    #[error("Unable to flush data from local tx buffer")]
+    LocalTxBufFlush(std::io::Error),
+    #[error("No free local port available for new host inititated connection")]
+    NoFreeLocalPort,
+    #[error("Backend rx queue is empty")]
+    EmptyBackendRxQ,
+}
+
+impl std::convert::From<Error> for std::io::Error {
+    fn from(e: Error) -> Self {
+        std::io::Error::new(io::ErrorKind::Other, e)
+    }
+}
+
+#[derive(Parser, Debug)]
+#[clap(version, about, long_about = None)]
+pub(crate) struct VsockArgs {
+    /// Context identifier of the guest which uniquely identifies the device for its lifetime.
+    #[clap(long, default_value_t = 3)]
+    guest_cid: u64,
+
+    /// Unix socket to which a hypervisor conencts to and sets up the control path with the device.
+    #[clap(long)]
+    socket: String,
+
+    /// Unix socket to which a host-side application connects to.
+    #[clap(long)]
+    uds_path: String,
+}
+
+#[derive(Debug, Clone)]
+/// This structure is the public API through which an external program
+/// is allowed to configure the backend.
+pub(crate) struct VsockConfig {
+    guest_cid: u64,
+    socket: String,
+    uds_path: String,
+}
+
+impl VsockConfig {
+    /// Create a new instance of the VsockConfig struct, containing the
+    /// parameters to be fed into the vsock-backend server.
+    pub fn new(guest_cid: u64, socket: String, uds_path: String) -> Self {
+        Self {
+            guest_cid,
+            socket,
+            uds_path,
+        }
+    }
+
+    /// Return the guest's current CID.
+    pub fn get_guest_cid(&self) -> u64 {
+        self.guest_cid
+    }
+
+    /// Return the path of the unix domain socket which is listening to
+    /// requests from the host side application.
+    pub fn get_uds_path(&self) -> String {
+        String::from(&self.uds_path)
+    }
+
+    /// Return the path of the unix domain socket which is listening to
+    /// requests from the guest.
+    pub fn get_socket_path(&self) -> String {
+        String::from(&self.socket)
+    }
+}
+
+impl TryFrom<VsockArgs> for VsockConfig {
+    type Error = Error;
+
+    fn try_from(cmd_args: VsockArgs) -> Result<Self> {
+        let socket = cmd_args.socket.trim().to_string();
+        let uds_path = cmd_args.uds_path.trim().to_string();
+
+        Ok(VsockConfig::new(cmd_args.guest_cid, socket, uds_path))
+    }
+}
+
+/// A local port and peer port pair used to retrieve
+/// the corresponding connection.
+#[derive(Hash, PartialEq, Eq, Debug, Clone)]
+pub struct ConnMapKey {
+    local_port: u32,
+    peer_port: u32,
+}
+
+impl ConnMapKey {
+    pub fn new(local_port: u32, peer_port: u32) -> Self {
+        Self {
+            local_port,
+            peer_port,
+        }
+    }
+}
+
+pub struct VhostUserVsockBackend {
+    guest_cid: __u64,
```

Comment:
Why __u64 here ?

Since we need to set this only once and then just send it in Little endian format, we can make it vm_memory::Le64 instead and set it as:

guest_cid = From::from(u64-value);

This is what's done in GPIO crate.

Reply:
yep, make sense to me!